src/net.h

```diff
@@ -1257,6 +1276,40 @@ class CConnman
      */
     std::vector<CService> m_onion_binds;
 
+    /**
+     * RAII helper to atomically create a copy of `vNodes` and add a reference
+     * to each of the nodes. The nodes are released when this object is destroyed.
+     */
+    class NodesSnapshot
+    {
+    public:
+        explicit NodesSnapshot(const CConnman& connman) : m_connman{connman}
+        {
+            LOCK(m_connman.cs_vNodes);
+            m_nodes_copy = m_connman.vNodes;
+            for (auto& node : m_nodes_copy) {
+                node->AddRef();
+            }
+        }
+
+        ~NodesSnapshot()
+        {
+            LOCK(m_connman.cs_vNodes);
+            for (auto& node : m_nodes_copy) {
+                node->Release();
+            }
+        }
+
+        const std::vector<CNode*>& Nodes() const
+        {
+            return m_nodes_copy;
+        }
+
+    private:
+        const CConnman& m_connman;
+        std::vector<CNode*> m_nodes_copy;
```

Comment:
```suggestion
        const std::vector<CNode*> m_nodes_copy;
```

Can `const` this if you add the constructor assignment to the initializer list (done in diff above)

Reply:
The copying and `AddRef()` should be done under the lock.

Reply:
Why `AddRef`? @amitiuttarwar suggestion can be accomplished with
```diff
         explicit NodesSnapshot(const CConnman& connman)
+            : m_nodes_copy(WITH_LOCK(connman.cs_vNodes, return connman.vNodes))
         {
-            LOCK(connman.cs_vNodes);
-            m_nodes_copy = connman.vNodes;
             for (auto& node : m_nodes_copy) {
                 node->AddRef();
             }
@@ -1305,7 +1304,7 @@ private:
         }

     private:
-        std::vector<CNode*> m_nodes_copy;
+        const std::vector<CNode*> m_nodes_copy;
     };
```

Reply:
Yeah :) `AddRef()` needs to be under the same lock that did the copy (without release in between) because otherwise some of the copied elements may be destroyed in the meantime and `node->` to try to dereference a dangling pointer while trying to call `AddRef()`.

It should work with something like
```cpp
explicit NodesSnapshot(const CConnman& connman)
            : m_nodes_copy(WITH_LOCK(connman.cs_vNodes, for (auto& node : connman.vNodes) { node->AddRef(); } ; return connman.vNodes))
```

but that is not worth it indeed :)